processing/src/main/java/io/druid/query/extraction/LookupReferencesManager.java

```diff
@@ -0,0 +1,183 @@
+  {
+    synchronized (lock) {
+      assertStarted();
+      if (lookupMap.containsKey(lookupName)) {
+        LOGGER.warn("lookup [%s] is not add, another lookup with the same name already exist", lookupName);
+        return false;
+      }
+      if (!lookupExtractorFactory.start()) {
+        throw new ISE("start method returned false for lookup [%s]", lookupName);
+      }
+      return (null == lookupMap.putIfAbsent(lookupName, lookupExtractorFactory));
+    }
+  }
+
+  /**
+   * @param lookups {@link Map<String, LookupExtractorFactory>} containing all the lookup as one batch.
+   *
+   * @throws IllegalStateException if the manager is closed or if {@link LookupExtractorFactory#start()} returns false
+   */
+  public void put(Map<String, LookupExtractorFactory> lookups)
+  {
+    Map<String, LookupExtractorFactory> faildExtractorFactoryMap = new HashMap<>();
+    synchronized (lock) {
+      assertStarted();
+      for (Map.Entry<String, LookupExtractorFactory> entry : lookups.entrySet()) {
+        final String lookupName = entry.getKey();
+        final LookupExtractorFactory lookupExtractorFactory = entry.getValue();
+        if (lookupMap.containsKey(lookupName)) {
+          LOGGER.warn("lookup [%s] is not add, another lookup with the same name already exist", lookupName);
+          continue;
+        }
+        if (!lookupExtractorFactory.start()) {
+          faildExtractorFactoryMap.put(lookupName, lookupExtractorFactory);
+          continue;
+        }
+        lookupMap.put(lookupName, lookupExtractorFactory);
+      }
+      if (!faildExtractorFactoryMap.isEmpty()) {
+        throw new ISE(
+            "was not able to start the following lookup(s) [%s]",
+            faildExtractorFactoryMap.keySet().toString()
+        );
+      }
+    }
+  }
+
+  /**
+   * @param lookupName name of {@link LookupExtractorFactory} to delete from the reference registry.
+   *                   this function does call the cleaning method {@link LookupExtractorFactory#close()}
+   *
+   * @return true only if {@code lookupName} is removed and the lookup correctly stopped
+   */
+  public boolean remove(String lookupName)
+  {
+    final LookupExtractorFactory lookupExtractorFactory = lookupMap.remove(lookupName);
+    if (lookupExtractorFactory != null) {
+      LOGGER.debug("Removing lookup [%s]", lookupName);
+      return lookupExtractorFactory.close();
+    }
+    return false;
+  }
+
+  /**
+   * @param lookupName key to fetch the reference of the object {@link LookupExtractor}
+   *
+   * @return reference of {@link LookupExtractorFactory} that correspond the {@code lookupName} or null if absent
+   *
+   * @throws IllegalStateException if the {@link LookupReferencesManager} is closed or did not start yet
+   */
+  @Nullable
+  public LookupExtractorFactory get(String lookupName)
+  {
+    final LookupExtractorFactory lookupExtractorFactory = lookupMap.get(lookupName);
+    assertStarted();
```

Comment:
Or a IAE or similar on not found lookup, when really it is just that we are either shutdown or not started.